provider/provider.go

```diff
@@ -0,0 +1,70 @@
+// Package provider implements structures and methods to provide blocks,
+// keep track of which blocks are provided, and to allow those blocks to
+// be reprovided.
+package provider
+
+import (
+	"context"
+
+	cid "github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log"
+	routing "github.com/libp2p/go-libp2p-routing"
+)
+
+var log = logging.Logger("provider")
+
+const provideOutgoingWorkerLimit = 8
+
+// Provider announces blocks to the network
+type Provider interface {
+	Run()
+	Provide(cid.Cid) error
+}
+
+type provider struct {
+	ctx context.Context
+	// the CIDs for which provide announcements should be made
+	queue *Queue
+	// used to announce providing to the network
+	contentRouting routing.ContentRouting
+}
+
+// NewProvider creates a provider that announces blocks to the network using a content router
+func NewProvider(ctx context.Context, queue *Queue, contentRouting routing.ContentRouting) Provider {
+	return &provider{
+		ctx:            ctx,
+		queue:          queue,
+		contentRouting: contentRouting,
+	}
+}
+
+// Start workers to handle provide requests.
+func (p *provider) Run() {
+	p.handleAnnouncements()
+}
+
+// Provide the given cid using specified strategy.
+func (p *provider) Provide(root cid.Cid) error {
+	p.queue.Enqueue(root)
+	return nil
+}
+
+// Handle all outgoing cids by providing (announcing) them
+func (p *provider) handleAnnouncements() {
+	for workers := 0; workers < provideOutgoingWorkerLimit; workers++ {
+		go func() {
+			for {
```

Comment:
Nit: This should probably check `p.ctx.Err() != nil`. Unfortunately, `select` selects from it's channels _randomly_ so this could keep dequeueing, even on shutdown.

(probably not going to happen but still something we should get better at doing, this is a "do what I say, not what I do" kind of comment).

Reply:
Is this what you're talking about? doing the check at the beginning of the loop and returning if we have an error?

Reply:
Ah, sorry. I literally just meant:

```
for p.ctx.Err() != nil { /* ... */ }
```

The context being canceled isn't an error or anything, it just means we need to stop (so we shouldn't be printing a warning). I only suggested this because the select statement may repeatedly select the dequeue case if both that and the `ctx.Done()` case are ready at the same time.